// @amplitude/session-replay-browser/lib/cjs/utils/rrweb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWindowWidth = exports.getWindowHeight = exports.getWindowScroll = void 0;
var tslib_1 = require("tslib");
var analytics_core_1 = require("@amplitude/analytics-core");
var rrweb_utils_1 = tslib_1.__importDefault(require("@amplitude/rrweb-utils"));
// These functions are not exposed in rrweb package, so we will define it here to use
// Ignoring this function since this is copied from rrweb
/* istanbul ignore next */
function getWindowScroll(win) {
    var _a, _b, _c, _d;
    var doc = win.document;
    return {
        left: doc.scrollingElement
            ? doc.scrollingElement.scrollLeft
            : win.pageXOffset !== undefined
                ? win.pageXOffset
                : doc.documentElement.scrollLeft ||
                    ((doc === null || doc === void 0 ? void 0 : doc.body) && ((_a = rrweb_utils_1.default.parentElement(doc.body)) === null || _a === void 0 ? void 0 : _a.scrollLeft)) ||
                    ((_b = doc === null || doc === void 0 ? void 0 : doc.body) === null || _b === void 0 ? void 0 : _b.scrollLeft) ||
                    0,
        top: doc.scrollingElement
            ? doc.scrollingElement.scrollTop
            : win.pageYOffset !== undefined
                ? win.pageYOffset
                : (doc === null || doc === void 0 ? void 0 : doc.documentElement.scrollTop) ||
                    ((doc === null || doc === void 0 ? void 0 : doc.body) && ((_c = rrweb_utils_1.default.parentElement(doc.body)) === null || _c === void 0 ? void 0 : _c.scrollTop)) ||
                    ((_d = doc === null || doc === void 0 ? void 0 : doc.body) === null || _d === void 0 ? void 0 : _d.scrollTop) ||
                    0,
    };
}
exports.getWindowScroll = getWindowScroll;
function getWindowHeight() {
    var globalScope = (0, analytics_core_1.getGlobalScope)();
    return ((globalScope === null || globalScope === void 0 ? void 0 : globalScope.innerHeight) ||
        (document.documentElement && document.documentElement.clientHeight) ||
        (document.body && document.body.clientHeight) ||
        0);
}
exports.getWindowHeight = getWindowHeight;
function getWindowWidth() {
    var globalScope = (0, analytics_core_1.getGlobalScope)();
    return ((globalScope === null || globalScope === void 0 ? void 0 : globalScope.innerWidth) ||
        (document.documentElement && document.documentElement.clientWidth) ||
        (document.body && document.body.clientWidth) ||
        0);
}
exports.getWindowWidth = getWindowWidth;
