// @amplitude/analytics-core/lib/esm/types/amplitude-context.js
import { RemoteConfigClient } from '../remote-config/remote-config';
import { Logger } from '../logger';
import { LogLevel } from './loglevel';
import { DEFAULT_INSTANCE_NAME } from './constants';
/**
 * @experimental
 * AmplitudeContext holds the core configuration and dependencies for an Amplitude instance.
 * This includes API key, instance name, server zone, remote config client, and logger.
 */
var AmplitudeContext = /** @class */ (function () {
    function AmplitudeContext(apiKey, instanceName, serverZone, logger) {
        if (instanceName === void 0) { instanceName = DEFAULT_INSTANCE_NAME; }
        if (serverZone === void 0) { serverZone = 'US'; }
        this.apiKey = apiKey;
        this.instanceName = instanceName;
        this.serverZone = serverZone;
        if (!logger) {
            var defaultLogger = new Logger();
            defaultLogger.enable(LogLevel.Error);
            this.logger = defaultLogger;
        }
        else {
            this.logger = logger;
        }
        this.remoteConfigClient = new RemoteConfigClient(this.apiKey, this.logger, this.serverZone);
    }
    return AmplitudeContext;
}());
export { AmplitudeContext };
