// @amplitude/plugin-experiment-browser/lib/cjs/experiment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.experimentPlugin = exports.ExperimentPlugin = void 0;
var tslib_1 = require("tslib");
var experiment_js_client_1 = require("@amplitude/experiment-js-client");
var ExperimentPlugin = /** @class */ (function () {
    function ExperimentPlugin(config) {
        this.name = ExperimentPlugin.pluginName;
        this.config = config;
    }
    ExperimentPlugin.prototype.setup = function (config, _client) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.experiment = (0, experiment_js_client_1.initializeWithAmplitudeAnalytics)(config.apiKey, this.config);
                return [2 /*return*/];
            });
        });
    };
    ExperimentPlugin.pluginName = '@amplitude/experiment-analytics-plugin';
    return ExperimentPlugin;
}());
exports.ExperimentPlugin = ExperimentPlugin;
var experimentPlugin = function (config) {
    return new ExperimentPlugin(config);
};
exports.experimentPlugin = experimentPlugin;
