// @amplitude/analytics-core/lib/cjs/types/amplitude-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplitudeContext = void 0;
var remote_config_1 = require("../remote-config/remote-config");
var logger_1 = require("../logger");
var loglevel_1 = require("./loglevel");
var constants_1 = require("./constants");
/**
 * @experimental
 * AmplitudeContext holds the core configuration and dependencies for an Amplitude instance.
 * This includes API key, instance name, server zone, remote config client, and logger.
 */
var AmplitudeContext = /** @class */ (function () {
    function AmplitudeContext(apiKey, instanceName, serverZone, logger) {
        if (instanceName === void 0) { instanceName = constants_1.DEFAULT_INSTANCE_NAME; }
        if (serverZone === void 0) { serverZone = 'US'; }
        this.apiKey = apiKey;
        this.instanceName = instanceName;
        this.serverZone = serverZone;
        if (!logger) {
            var defaultLogger = new logger_1.Logger();
            defaultLogger.enable(loglevel_1.LogLevel.Error);
            this.logger = defaultLogger;
        }
        else {
            this.logger = logger;
        }
        this.remoteConfigClient = new remote_config_1.RemoteConfigClient(this.apiKey, this.logger, this.serverZone);
    }
    return AmplitudeContext;
}());
exports.AmplitudeContext = AmplitudeContext;
